### DISCUSSION TITLE
Typing error: Incompatible types in assignment (expression has type "ConfigDict", base class "SQLModel" defined the type as "SQLModelConfig")

### DISCUSSION THREAD

**Discussion by copdips** [fastapi/sqlmodel#855] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from pydantic import ConfigDict

class HeroCreate(SQLModel):
    name: str | None = None

    class Config:
        extra = "forbid"

class HeroUpdate(SQLModel):
    name: str | None = None
    model_config = ConfigDict(extra="forbid")
```

### Description

Hello,

I need to add a config to prevent from extra fields when creating or updating Hero model.

Above example works well, but my VSCode raises linting errors

When using:

```python
    class Config:
        extra = "forbid"
```

The linting error thrown by `Pylance` is:

```console
"Config" overrides symbol of same name in class "SQLModel"
  "app.db.models.HeroCreate.Config" is incompatible with "sqlmodel.main.SQLModel.Config"
  Type "type[app.db.models.HeroCreate.Config]" cannot be assigned to type "type[sqlmodel.main.SQLModel.Config]"
```

When using the second format:

```python
model_config: ConfigDict = ConfigDict(extra="forbid")
````

The linting error thrown by `MyPy` is:

```console
Incompatible types in assignment (expression has type "ConfigDict", base class "SQLModel" defined the type as "SQLModelConfig")
```

### Operating System

Linux

### SQLModel Version

lastest: 0.0.16

### Python Version

3.11.7

### Additional Context

mypy                 1.9.0
mypy-extensions      1.0.0

**Comment by SantiMNV** [comment]
I experience the same, any updates on this?

**Reply by joachimhuet** [reply]
It looks like `SQLModelConfig` still inherits the deprecated `BaseConfig` class. `sqlmodel._compat.py`.
If you change `BaseConfig` parent class with `pydantic.ConfigDict` does your warning remains ?

**Reply by copdips** [reply]
It seems that it uses already `pydantic.ConfigDict`:
https://github.com/tiangolo/sqlmodel/blob/6151f23e15703223128d641081b63d49a658d524/sqlmodel/_compat.py#L72-L73

**Reply by joachimhuet** [reply]
Oh yes you are right, it seems my editor point to the `else` condition by default. Probably mypy does not know where to look at neither.

I tried to reproduce and i get rid of the warning by doing that: 
```python
from sqlmodel._compat import SQLModelConfig

class BaseModel(SQLModel):
    model_config = cast(SQLModelConfig, {"extra": "forbid"})

```

But I'm not sure it is a good idea, just putting a `# type: ignore` would also safely work by passing a dict to `model_config`

```python
class BaseModel(SQLModel):
    model_config = {"extra": "forbid"}  # type: ignore
```

**Reply by copdips** [reply]
if we were to accept `# type: ignore`, I would then add it directly to `model_config = ConfigDict(extra="forbid")`

**Reply by zwlxt** [reply]
My workaround is creating a BaseModel which inherits your customized pydantic.BaseModel and SQLModel. It seems to satisfy pylance (in strict mode).

I am using sqlmodel==0.0.24 and Python 3.13.3 

```python
class CamelCasedModel(pydantic.BaseModel):
    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
    )

# note: first CamelCasedModel then SQLModel
class BaseModel(CamelCasedModel, SQLModel):
    pass

# then use it in your entity class
class UserAccount(BaseModel, table=True):
    ...
```

**Comment by natehouk** [comment]
Any update on this?

**Reply by aemonge** [reply]
No update?

**Comment by bartelemi** [comment]
Seems like this is still a problem even on v0.0.23 and v0.0.24 of sqlmodel and pydantic v2.
There were some small changes in how the IS_PYDANTIC_V2 typing guard is constructed, but mypy is still struggling to understand it and defaults to the `else:` statement (thinking that the code is running pydantic v1).

**Reply by onyx093** [reply]
Instead of writing it like this

```
from pydantic import ConfigDict

class HeroCreate(SQLModel):
    name: str | None = None

class HeroUpdate(SQLModel):
    name: str | None = None
    model_config = ConfigDict(from_attributes=True)
```

just write it like this

```
class HeroCreate(SQLModel):
    name: str | None = None

    class Config:
        extra = "forbid"

class HeroUpdate(SQLModel, from_attributes=True):
    name: str | None = None
```
